Intermittent failure of DynamicCallResourceTest.testConcurrentDynamicCalls

### Describe the bug

`DynamicCallResourceTest.testConcurrentDynamicCalls` fails intermittently.

https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/nightly/job/kogito-runtimes.build-and-deploy/75/
https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/nightly/job/kogito-runtimes.build-and-deploy/77/

```
[2024-08-20T03:18:17.225Z] [ERROR] Tests run: 7, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 28.01 s <<< FAILURE! -- in org.kie.kogito.process.dynamic.DynamicCallResourceTest
[2024-08-20T03:18:17.225Z] [ERROR] org.kie.kogito.process.dynamic.DynamicCallResourceTest.testConcurrentDynamicCalls -- Time elapsed: 4.325 s <<< ERROR!
[2024-08-20T03:18:17.225Z] java.util.concurrent.ExecutionException:
[2024-08-20T03:18:17.225Z] java.lang.AssertionError: 1 expectation failed.
[2024-08-20T03:18:17.225Z] Expected status code <200> but was <400>.
[2024-08-20T03:18:17.225Z]
[2024-08-20T03:18:17.225Z]      at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
[2024-08-20T03:18:17.225Z]      at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
[2024-08-20T03:18:17.225Z]      at org.kie.kogito.process.dynamic.DynamicCallResourceTest.testConcurrentDynamicCalls(DynamicCallResourceTest.java:168)
[2024-08-20T03:18:17.225Z]      at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[2024-08-20T03:18:17.225Z]      at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:1013)
[2024-08-20T03:18:17.225Z]      at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:827)
[2024-08-20T03:18:17.225Z]      at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
[2024-08-20T03:18:17.225Z]      at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

### Expected behavior

Test passes constantly.

### Actual behavior

Test fails intermittently. we can confirm it in nightly build as above. Locally, I see it fails once in 4 or 5 times testing.

### How to Reproduce?

```
mvn clean install -Dquickly
cd quarkus/addons/dynamic/integration-tests
mvn test -Dtest=DynamicCallResourceTest
```
Locally, I see it fails once in 4 or 5 times testing.

### Output of `uname -a` or `ver`

Linux tkobayas-thinkpadt14sgen1.tokyo.csb 6.10.4-100.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Sun Aug 11 15:56:38 UTC 2024 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.12" 2024-07-16 OpenJDK Runtime Environment (Red_Hat-17.0.12.0.7-2) (build 17.0.12+7) OpenJDK 64-Bit Server VM (Red_Hat-17.0.12.0.7-2) (build 17.0.12+7, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Kogito version or git rev (or at least Quarkus version if you are using Kogito via Quarkus platform BOM)

main branch

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

_No response_

--- a/addons/common/process-dynamic/src/main/java/org/kie/kogito/process/dynamic/ProcessInstanceDynamicCallHelper.java
+++ b/addons/common/process-dynamic/src/main/java/org/kie/kogito/process/dynamic/ProcessInstanceDynamicCallHelper.java
@@ -18,9 +18,9 @@
  */
 package org.kie.kogito.process.dynamic;
 
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.stream.Stream;
 
 import org.drools.core.common.InternalKnowledgeRuntime;
 import org.jbpm.process.instance.InternalProcessRuntime;
@@ -41,8 +41,8 @@ public class ProcessInstanceDynamicCallHelper {
     private ProcessInstanceDynamicCallHelper() {
     }
 
-    public static void executeRestCall(RestWorkItemHandler handler, Stream<Process<?>> processes, String processId, String processInstanceId, RestCallInfo input) {
-        Process<?> processDef = processes.filter(p -> p.id().equals(processId)).findAny().orElseThrow(() -> new IllegalArgumentException("Cannot find process " + processId));
+    public static void executeRestCall(RestWorkItemHandler handler, Collection<Process<?>> processes, String processId, String processInstanceId, RestCallInfo input) {
+        Process<?> processDef = processes.stream().filter(p -> p.id().equals(processId)).findAny().orElseThrow(() -> new IllegalArgumentException("Cannot find process " + processId));
         WorkflowProcessInstance pi = (WorkflowProcessInstance) findProcessInstance(processDef, processInstanceId);
         WorkflowProcessImpl process = (WorkflowProcessImpl) pi.getProcess();
         if (!process.isDynamic()) {

--- a/jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/expr/ExpressionHandlerFactory.java
+++ b/jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/expr/ExpressionHandlerFactory.java
@@ -18,15 +18,18 @@
  */
 package org.kie.kogito.process.expr;
 
+import java.util.Collection;
 import java.util.Optional;
 import java.util.ServiceLoader;
+import java.util.stream.Collectors;
 
 public class ExpressionHandlerFactory {
 
     private ExpressionHandlerFactory() {
     }
 
-    private static final ServiceLoader<ExpressionHandler> serviceLoader = ServiceLoader.load(ExpressionHandler.class);
+    private static final Collection<ExpressionHandler> expressionHandlers = ServiceLoader.load(ExpressionHandler.class).stream()
+            .map(ServiceLoader.Provider::get).collect(Collectors.toUnmodifiableList());
 
     public static Expression get(String lang, String expr) {
         return getExpressionHandler(lang).orElseThrow(
@@ -34,10 +37,10 @@ public class ExpressionHandlerFactory {
     }
 
     public static boolean isSupported(String lang) {
-        return serviceLoader.stream().anyMatch(p -> p.get().lang().equals(lang));
+        return expressionHandlers.stream().map(ExpressionHandler::lang).anyMatch(lang::equals);
     }
 
     private static Optional<ExpressionHandler> getExpressionHandler(String lang) {
-        return serviceLoader.stream().filter(p -> p.get().lang().equals(lang)).findFirst().map(p -> p.get());
+        return expressionHandlers.stream().filter(p -> p.lang().equals(lang)).findFirst();
     }
 }

--- a/quarkus/addons/dynamic/runtime/src/main/java/org/kie/kogito/process/dynamic/ProcessInstanceDynamicCallsResource.java
+++ b/quarkus/addons/dynamic/runtime/src/main/java/org/kie/kogito/process/dynamic/ProcessInstanceDynamicCallsResource.java
@@ -18,6 +18,9 @@
  */
 package org.kie.kogito.process.dynamic;
 
+import java.util.Collection;
+import java.util.stream.Collectors;
+
 import org.kie.kogito.process.Process;
 import org.kogito.workitem.rest.RestWorkItemHandler;
 
@@ -40,12 +43,16 @@ import jakarta.ws.rs.core.Response;
 public class ProcessInstanceDynamicCallsResource {
 
     @Inject
-    Instance<Process<?>> processes;
-    @Inject
     Vertx vertx;
     @Inject
     WebClientOptions sslOptions;
     private RestWorkItemHandler handler;
+    private Collection<Process<?>> processes;
+
+    @Inject
+    ProcessInstanceDynamicCallsResource(Instance<Process<?>> processes) {
+        this.processes = processes.stream().collect(Collectors.toUnmodifiableList());
+    }
 
     @PostConstruct
     void init() {
@@ -57,8 +64,7 @@ public class ProcessInstanceDynamicCallsResource {
     @Consumes(MediaType.APPLICATION_JSON)
     @Path("/{processId}/{processInstanceId}/rest")
     public Response executeRestCall(@PathParam("processId") String processId, @PathParam("processInstanceId") String processInstanceId, RestCallInfo input) {
-        ProcessInstanceDynamicCallHelper.executeRestCall(handler, processes.stream(), processId, processInstanceId, input);
+        ProcessInstanceDynamicCallHelper.executeRestCall(handler, processes, processId, processInstanceId, input);
         return Response.status(200).build();
     }
-
 }

